// packages/shared/shared/grpc/protos/ocr.proto
syntax = "proto3";

package ocr;

import "common.proto";

// ============================================
// OCR 추출 서비스
// ============================================

service OCRService {
    // 단일 이미지 OCR 추출
    rpc ExtractText(OCRRequest) returns (OCRResponse);

    // 배치 이미지 OCR 추출 (Server Streaming)
    rpc ExtractTextBatch(OCRBatchRequest) returns (stream OCRBatchProgress);

    // 상태 확인
    rpc CheckHealth(HealthCheckRequest) returns (HealthCheckResponse);
}

// ============================================
// 요청/응답 메시지
// ============================================

// OCR 요청
message OCRRequest {
    string public_image_path = 1;
    string private_image_path = 2;
    string language = 3;                    // 기본값: "korean"
    float confidence_threshold = 4;         // 기본값: 0.5
    bool use_angle_cls = 5;                 // 기본값: true
    common.Metadata options = 6;            // 추가 옵션
}

// OCR 응답
message OCRResponse {
    common.Status status = 1;
    string text = 2;                        // 추출된 전체 텍스트
    float overall_confidence = 3;           // 전체 신뢰도
    repeated TextBox text_boxes = 4;        // 텍스트 박스 리스트
    common.Metadata metadata = 5;           // 엔진 정보, 처리 시간 등
    common.ErrorInfo error = 6;             // 에러 정보 (실패 시)
}

// 텍스트 박스
message TextBox {
    string text = 1;
    float confidence = 2;
    common.BoundingBox bbox = 3;
}

// ============================================
// 배치 처리
// ============================================

// 배치 요청
message OCRBatchRequest {
    repeated ImagePath image_paths = 1;
    string language = 2;
    float confidence_threshold = 3;
    bool use_angle_cls = 4;
}

// 이미지 경로
message ImagePath {
    string public_path = 1;
    string private_path = 2;
}

// 배치 진행 상황 (스트리밍)
message OCRBatchProgress {
    string batch_id = 1;
    int32 total_images = 2;
    int32 processed_images = 3;
    OCRResponse current_result = 4;         // 현재 처리 결과
    float progress_percentage = 5;
}

// ============================================
// 헬스 체크
// ============================================

message HealthCheckRequest {
    string service_name = 1;
}

message HealthCheckResponse {
    common.Status status = 1;
    string engine_type = 2;                 // "easyocr" | "paddleocr" | "mock"
    bool model_loaded = 3;
    string version = 4;
}
